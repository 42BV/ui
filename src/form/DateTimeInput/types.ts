// See https://momentjs.com/docs/#/displaying/format/ for formatting options.
export type DateFormat = string | false; // When false it should not be used
export type TimeFormat = string | false; // When false it should not be used
