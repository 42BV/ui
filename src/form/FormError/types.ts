export type OnChange = (hasErrors: boolean) => void;