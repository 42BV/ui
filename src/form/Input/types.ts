export type Position = 'left' | 'right';
export type Mask = (string | RegExp)[];
