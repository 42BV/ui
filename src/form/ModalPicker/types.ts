export type AddButtonCallback<T> = () => Promise<T>;

export interface AddButtonOptions<T> {
  label: string;
  onClick: AddButtonCallback<T>;
}
