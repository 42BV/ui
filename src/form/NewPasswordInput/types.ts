export type NewPasswordInputRule =
  | 'lowercase'
  | 'uppercase'
  | 'number'
  | 'specialChar'
  | 'minimumLength'
  | 'noSpace';
