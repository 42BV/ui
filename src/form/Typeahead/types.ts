/**
 * Represents the value of a Typeahead.
 */
export type TypeaheadOption<T> = {
  label: string;
  value: T;
};
