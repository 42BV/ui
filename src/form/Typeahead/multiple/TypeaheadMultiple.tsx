import React, { useState } from 'react';
import { FormGroup, Label } from 'reactstrap';
import {
  AsyncTypeahead,
  TokenProps,
  Typeahead,
  TypeaheadProps
} from 'react-bootstrap-typeahead';
import { TypeaheadOption } from '../types';
import { withJarb } from '../../withJarb/withJarb';
import { doBlur, alwaysTrue } from '../../utils';
import { optionToTypeaheadOption } from '../utils';
import {
  FieldCompatibleWithPredeterminedOptions,
  isOptionSelected
} from '../../option';
import classNames from 'classnames';
import { Tag } from '../../../core/Tag/Tag';
import { FieldCompatible } from '../../types';
import { useId } from '../../../hooks/useId/useId';
import { useOptions } from '../../useOptions';
import { t } from '../../../utilities/translation/translation';

type Text = {
  /**
   * Text to show when more than 100 items are available
   * in the dropdown. This is used by the pagination built
   * into the Typeahead of reactstrap.
   *
   * Defaults to `Display additional results...`
   */
  paginationText?: string;
};

type Props<T> = FieldCompatible<T[], T[] | undefined> &
  FieldCompatibleWithPredeterminedOptions<T> & {
    /**
     * Optionally specify the number of suggestions to fetch for the
     * dropdown.
     *
     * When `options` is an array, all options will always be used
     * and this prop has no effect.
     *
     * When `options` is a fetcher, it will determine how many options
     * are requested through the fetcher as the `page` parameter.
     * This means that when you set the pageSize to `100` that
     * `100` items are fetched from the back-end. Beware of
     * performance issues when setting the value too high.
     *
     * Beware that setting the page size too high will cause the UX
     * to deteriorate on smaller screens!
     *
     * Defaults to `10`.
     */
    pageSize?: number;

    /**
     * Optionally specify the number of suggestions to show in the
     * dropdown.
     *
     * When `options` is an array, this prop will limit the amount of
     * suggestions and display pagination.
     *
     * When `options` is a fetcher, this prop will limit the amount of
     * suggestions only when `pageSize` is larger than this prop. This
     * means the options fetched from the back-end might be a large list
     * but only a subset of them are displayed at the same time. Every
     * time the user types in the input, the options are fetched again.
     * Beware of performance issues when setting the `pageSize` too high.
     *
     * Defaults to `100`.
     */
    maxResults?: number;

    /**
     * Optionally customized text within the component.
     * This text should already be translated.
     */
    text?: Text;
  };

/**
 * The TypeaheadMultiple is a form element which allows the user
 * to select multiple options by searching for them and selecting
 * them.
 *
 * The use case is that when there are too many options to render
 * in a simple Select you can use the TypeaheadMultiple.
 *
 * Use the TypeaheadMultiple when the user knows which options he / she
 * wants to select beforehand. Scenario: you are building a system
 * for a trading company, the users need to enter some trading code
 * from a big list. The users know each code by heart since they
 * work with the system daily. This is a nice use case for the TypeaheadMultiple
 * because the user can type in faster than the can select from a
 * ModalPickerMultiple.
 */
export function TypeaheadMultiple<T>(props: Props<T>) {
  const {
    id,
    label,
    placeholder,
    value,
    color,
    error,
    labelForOption,
    onFocus,
    valid,
    className = '',
    onChange,
    onBlur,
    options,
    keyForOption,
    isOptionEqual,
    reloadOptions,
    isOptionEnabled = alwaysTrue,
    pageSize = 10,
    text = {},
    maxResults = 100
  } = props;

  const [query, setQuery] = useState('');

  const { page, loading } = useOptions<T>({
    options,
    value,
    keyForOption,
    isOptionEqual,
    labelForOption,
    reloadOptions,
    query,
    pageNumber: 1,
    size: Array.isArray(options) ? options.length : pageSize,
    optionsShouldAlwaysContainValue: false
  });

  const typeaheadOptions = page.content
    .filter((option) => isOptionEnabled(option))
    .filter(
      (option) =>
        !isOptionSelected({
          option,
          keyForOption,
          labelForOption,
          isOptionEqual,
          value
        })
    )
    .map((option) => optionToTypeaheadOption(option, labelForOption));

  function doOnChange(values: TypeaheadOption<T>[]): void {
    if (values.length === 0) {
      onChange(undefined);
    } else {
      onChange(values.map((option) => option.value));
    }

    // Due this: https://github.com/ericgio/react-bootstrap-typeahead/issues/224
    // If we invoke the input.onBlur at the typeahead onblur you would get this:
    // onBlur -> onChange, but it should be the other way around.
    // onBlur Should be called when the user navigates away from the input.
    // In this case when the user selects an item (onChange).
    doBlur(onBlur);
  }

  function renderToken(
    option: TypeaheadOption<T>,
    props: TokenProps,
    index: number
  ) {
    return (
      <Tag
        key={index}
        text={option.label}
        // @ts-expect-error The prop onRemove actually exists, the typings are wrong
        onRemove={() => props.onRemove(option)}
        className="align-self-center"
      />
    );
  }

  let selected: TypeaheadOption<T>[] = [];
  if (value && value.length) {
    selected = value.map((v) => optionToTypeaheadOption(v, labelForOption));
  }

  const classes = classNames(className, {
    'is-invalid': valid === false
  });

  const innerId = useId({ id });

  const typeaheadProps: TypeaheadProps<TypeaheadOption<T>> = {
    id,
    filterBy: alwaysTrue,
    paginationText: t({
      key: 'TypeaheadMultiple.PAGINATION_TEXT',
      fallback: 'Display additional results...',
      overrideText: text.paginationText
    }),
    multiple: true,
    placeholder,
    selected,
    options: typeaheadOptions,
    onChange: doOnChange,
    onFocus,
    inputProps: {
      id: innerId,
      // @ts-expect-error The input props value works
      value,
      className: classNames('form-control', {
        'is-invalid': valid === false
      })
    },
    renderToken,
    maxResults,
    paginate: true
  };

  return (
    <FormGroup className={classes} color={color}>
      {label ? <Label for={innerId}>{label}</Label> : null}
      <div className={selected.length === 0 ? 'showing-placeholder' : ''}>
        {Array.isArray(options) ? (
          <Typeahead {...typeaheadProps} onInputChange={setQuery} />
        ) : (
          <AsyncTypeahead
            {...typeaheadProps}
            isLoading={loading}
            delay={200}
            onSearch={setQuery}
          />
        )}
      </div>
      {error}
    </FormGroup>
  );
}

/**
 * Variant of the TypeaheadMultiple which can be used in a Jarb context.
 */
export const JarbTypeaheadMultiple = withJarb<
  any[],
  any[] | undefined,
  Props<any>
>(TypeaheadMultiple);
