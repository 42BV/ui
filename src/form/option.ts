import { Page } from '@42.nl/spring-connect';

export type Options<T> = FetchOptionsCallback<T> | T[];

/**
 * Variant of field compatible where the options the user can choose
 * are predetermined.
 */
export type FieldCompatibleWithPredeterminedOptions<T> = {
  /**
   * Is either a callback to fetch the options to display to the user.
   * When options is a callback it will not execute when the callback
   * changes, only when the `reloadOptions` changes will the callback
   * be executed again. This means that it is safe to pass in a
   * lambda / anonymous / unstable function here.
   *
   * Or an array of fixed options.
   */
  options: Options<T>;

  /**
   * Callback to convert a value of type T to an option to show
   * to the user.
   */
  labelForOption: LabelForOption<T>;

  /**
   * Optional callback which is used to determine if two options
   * of type T are equal.
   *
   * When `isOptionEqual` is not defined the outcome of `labelForOption`
   * is used to test equality.
   */
  isOptionEqual?: IsOptionEqual<T>;

  /**
   * Optional callback to get a unique key for an option.
   * This is used to provide each option in the form element a unique key.
   * Defaults to the 'id' property if it exists, otherwise uses labelForOption.
   */
  keyForOption?: KeyForOption<T>;

  /**
   * Optional callback which is called for every option to determine
   * if the option can be selected. By default, all options can be
   * selected.
   */
  isOptionEnabled?: IsOptionEnabled<T>;

  /**
   * Optionally a value to detect changes and trigger the
   * `options` to reload the options, by fetching them again.
   *
   * Whenever the value of `reloadOptions` changes the options are
   * reloaded. This gives the developer an external way to trigger
   * the reloading of the options.
   */
  reloadOptions?: string | number | boolean | null | undefined;
};

/**
 * Callback to determine the label for a given value of type T.
 * Aka the text the user sees when selecting a value.
 */
export type LabelForOption<T> = (option: T) => string;

/**
 * Callback to determine the unique identifier for a given value of type T.
 */
export type KeyForOption<T> = (value: T) => React.Key;

/**
 * Callback to determine if two options are equal to each other
 */
export type IsOptionEqual<T> = (a: T, b: T) => boolean;

/**
 * Callback to determine if the option is currently enabled.
 */
export type IsOptionEnabled<T> = (option: T) => boolean;

export type FetchOptionsCallbackConfig = {
  /**
   * The query which was provided by the component. If no query is
   * present an empty string is provided.
   */
  query: string;

  /**
   * The page number the developer must fetch from the back-end.
   * Starts at 1.
   */
  page: number;

  /**
   * The desired size which the form component accepts. The developer
   * should not provide more items than the size, otherwise unintended
   * things may happen.
   */
  size: number;
};

/**
 * Callback which should return a Page of options which can
 * be selected by the user.
 *
 * It is given three parameters:
 *
 * 1. `query` A string you must use to filter the number of options.
 * Used for searches.
 *
 * 2. `page` A number telling you which page you must load. Used
 * to limit the total number of elements by asking for a small slice
 * each time.
 *
 * 3. `size` A number telling you the size of the Page you must load.
 * Often components can only render so much options before becoming
 * unwieldy. They can tell you the size of options they support
 * through this parameter.
 */
export type FetchOptionsCallback<T> = (
  config: FetchOptionsCallbackConfig
) => Promise<Page<T>>;

type KeyForOptionConfig<T> = {
  option: T;
  keyForOption?: KeyForOption<T>;
  labelForOption: LabelForOption<T>;
};

/**
 * Returns a unique key for an option.
 *
 * Will try the following ways to get a unique key in this order:
 *
 * 1. Will use: `keyForOption` if provided. In this case we
 *    must trust the developer to implement it correctly.
 *
 * 2. If the options has an `id` we expect that id to be unique and
 *    that `id` is used as the key.
 *
 * 3. As a last resort we will use the `labelForOption` in this case
 *    the label of the option will become the key.
 */
export function getKeyForOption<T>({
  option,
  keyForOption,
  labelForOption
}: KeyForOptionConfig<T>) {
  if (keyForOption) {
    return keyForOption(option);
  }

  // @ts-expect-error Accept that the option could have an id
  if (option.id) {
    // @ts-expect-error Accept that the option could have an id
    return `${option.id}`;
  }

  return labelForOption(option);
}

type IsOptionSelectedConfig<T> = {
  option: T;
  keyForOption?: KeyForOption<T>;
  labelForOption: LabelForOption<T>;
  isOptionEqual?: IsOptionEqual<T>;
  value?: T[] | T;
};

/**
 * Determines if is the option is selected given the value.
 *
 * Will prefer to use the `isOptionEqual` to check if the value is
 * selected, otherwise it will fall back to checking the key.
 *
 * Note: always pass along every argument even the optional ones.
 * so `isOptionSelected` can perform its function best.
 */
export function isOptionSelected<T>(
  config: IsOptionSelectedConfig<T>
): boolean {
  const { value } = config;

  if (!value) {
    return false;
  }

  if (Array.isArray(value)) {
    return value.some((value) => isValueSelected(value, config));
  } else {
    return isValueSelected(value, config);
  }
}

// Helper for isOptionSelected
function isValueSelected<T>(
  value: T,
  config: IsOptionSelectedConfig<T>
): boolean {
  const { option, isOptionEqual, labelForOption, keyForOption } = config;

  if (isOptionEqual) {
    return isOptionEqual(value, option);
  } else {
    const key = getKeyForOption({ option, keyForOption, labelForOption });
    return (
      key === getKeyForOption({ option: value, keyForOption, labelForOption })
    );
  }
}
