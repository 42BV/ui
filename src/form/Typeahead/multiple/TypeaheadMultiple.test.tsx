import React from 'react';
import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom';

import { TypeaheadMultiple } from './TypeaheadMultiple';

import { adminUser, listOfUsers, pageOfUsersFetcher, userUser } from '../../../test/fixtures';
import { User } from '../../../test/types';

import { pageOf } from '../../../utilities/page/page';
import { useOptions } from '../../useOptions';
import { IsOptionEnabled } from '../../option';
import userEvent from '@testing-library/user-event';

jest.mock('../../useOptions', () => {
  return { useOptions: jest.fn() };
});

describe('Component: TypeaheadMultiple', () => {
  function setup({
    value,
    hasPlaceholder,
    hasLabel,
    hasId = true,
    loading = false,
    isOptionEnabled,
    isAsync,
    pageSize,
    maxResults,
    allowNew,
    text
  }: {
    value?: User[];
    hasPlaceholder?: boolean;
    hasLabel?: boolean;
    hasId?: boolean;
    loading?: boolean;
    isOptionEnabled?: IsOptionEnabled<User>;
    isAsync?: boolean;
    pageSize?: number;
    maxResults?: number;
    allowNew?: true;
    text?: {
      paginationText?: string;
    };
  }) {
    const onChangeSpy = jest.fn();
    const onBlurSpy = jest.fn();

    // @ts-expect-error This is in fact a mock
    useOptions.mockImplementation(
      ({ pageNumber, size, optionsShouldAlwaysContainValue }) => {
        expect(pageNumber).toBe(1);
        expect(size).toBe(!isAsync ? 3 : pageSize ?? 10);
        expect(optionsShouldAlwaysContainValue).toBe(false);

        return {
          page: pageOf(listOfUsers(), pageNumber, size),
          loading
        };
      }
    );

    const props = {
      placeholder: hasPlaceholder ? 'Please provide your best friend' : undefined,
      options: isAsync ? pageOfUsersFetcher : listOfUsers(),
      labelForOption: (user: User) => user.email,
      isOptionEnabled,
      value,
      onChange: onChangeSpy,
      onBlur: onBlurSpy,
      pageSize,
      maxResults,
      allowNew,
      text,
      error: 'Some error',
      id: hasId ? 'bestFriend' : undefined,
      label: 'Best friend',
      hiddenLabel: !hasLabel
    };

    const { container, rerender, asFragment } = render(
      <TypeaheadMultiple {...props} />
    );

    return { container, props, rerender, asFragment, onChangeSpy, onBlurSpy };
  }

  describe('ui', () => {
    test('default', () => {
      const { container } = setup({});
      expect(container).toMatchSnapshot();
    });

    test('with value', () => {
      setup({ value: [ adminUser() ] });
      expect(screen.queryAllByText('×').length).toBe(1);
      expect(screen.getByText('×').parentNode?.textContent).toBe('admin@42.nl×');
    });

    test('with placeholder', () => {
      setup({ hasPlaceholder: true });
      expect(screen.queryByPlaceholderText('Please provide your best friend')).toBeInTheDocument();
    });

    test('without placeholder', () => {
      setup({ hasPlaceholder: false });
      expect(screen.queryByPlaceholderText('Please provide your best friend')).not.toBeInTheDocument();
    });

    test('without id', () => {
      setup({ hasId: false, hasLabel: true });
      expect(screen.queryByLabelText('Best friend')).toBeInTheDocument();
    });

    test('visible label', () => {
      setup({ hasLabel: true });
      expect(screen.queryByText('Best friend')).toBeInTheDocument();
      expect(screen.queryByLabelText('Best friend')).toBeInTheDocument();
    });

    test('invisible label', () => {
      setup({ hasLabel: false });
      expect(screen.queryByText('Best friend')).not.toBeInTheDocument();
      expect(screen.queryByLabelText('Best friend')).toBeInTheDocument();
    });

    test('async with a custom pageSize of 2 options in the dropdown', async () => {
      expect.assertions(7);

      const { asFragment } = setup({
        isAsync: true,
        pageSize: 2
      });

      expect(useOptions).toBeCalledTimes(1);
      expect(useOptions).toBeCalledWith(expect.objectContaining({ size: 2 }));

      fireEvent.change(screen.getByRole('textbox'), { target: { value: '42.nl' } });

      await waitFor(() => {
        expect(screen.queryAllByText('42.nl').length).toBe(2);
      });
      expect(asFragment()).toMatchSnapshot();
    });

    test('async with the default pageSize of 10 options in the dropdown', async () => {
      expect.assertions(7);

      const { asFragment } = setup({ isAsync: true });

      expect(useOptions).toBeCalledTimes(1);
      expect(useOptions).toBeCalledWith(expect.objectContaining({ size: 10 }));

      fireEvent.change(screen.getByRole('textbox'), { target: { value: '42.nl' } });

      await waitFor(() => {
        expect(screen.queryAllByText('42.nl').length).toBe(3);
      });
      expect(asFragment()).toMatchSnapshot();
    });

    test('with the default pagination text', async () => {
      expect.assertions(5);

      const { asFragment } = setup({
        isAsync: true,
        maxResults: 2
      });

      fireEvent.change(screen.getByRole('textbox'), { target: { value: '42.nl' } });

      await waitFor(() => {
        expect(screen.queryByText('Display additional results...')).toBeInTheDocument();
      });
      expect(asFragment()).toMatchSnapshot();
    });

    test('with a custom pagination text', async () => {
      expect.assertions(6);

      const { asFragment } = setup({
        isAsync: true,
        maxResults: 2,
        text: {
          paginationText: 'Show more'
        }
      });

      fireEvent.change(screen.getByRole('textbox'), { target: { value: '42.nl' } });

      await waitFor(() => {
        expect(screen.queryByText('Display additional results...')).not.toBeInTheDocument();
        expect(screen.queryByText('Show more')).toBeInTheDocument();
      });
      expect(asFragment()).toMatchSnapshot();
    });

    test('loading', () => {
      const { container } = setup({ loading: true, isAsync: true });
      expect(screen.queryByText('Loading...')).toBeInTheDocument();
      expect(container).toMatchSnapshot();
    });
  });

  describe('renderToken', () => {
    it('should when the Tag is closed call onRemove', () => {
      const { onChangeSpy } = setup({
        value: [ adminUser() ]
      });

      fireEvent.click(screen.getByText('×'));

      expect(onChangeSpy).toBeCalledTimes(1);
      expect(onChangeSpy).toBeCalledWith(undefined);
    });
  });

  describe('events', () => {
    test('onChange', async () => {
      expect.assertions(21);

      const { onBlurSpy, onChangeSpy } = setup({
        value: []
      });

      await userEvent.type(screen.getByRole('textbox'), 'admin');
      fireEvent.click(screen.getByLabelText('admin@42.nl'));

      expect(onChangeSpy).toHaveBeenCalledTimes(1);
      expect(onChangeSpy).toHaveBeenCalledWith([ adminUser() ]);
      expect(onBlurSpy).toHaveBeenCalledTimes(1);
    });

    it('should set the query when the user starts typing in the input field', async () => {
      expect.assertions(16);
      setup({
        isAsync: true
      });

      expect(useOptions).toBeCalledTimes(1);

      fireEvent.change(screen.getByRole('textbox'), { target: { value: 'admin' } });

      await waitFor(() => {
        expect(useOptions).toBeCalledTimes(2);
        expect(useOptions).toBeCalledWith(
          expect.objectContaining({ query: 'admin' })
        );
      });
    });

    it('should filter out already selected values and disabled options from the typeahead options', async () => {
      expect.assertions(6);

      setup({
        value: [ adminUser() ], // The admin user is select, so it should be filtered out
        isOptionEnabled: (user) => user.id !== userUser().id // Also disable the userUser
      });

      fireEvent.focus(screen.getByRole('textbox'));

      await waitFor(() => {
        expect(screen.queryByLabelText('admin@42.nl')).not.toBeInTheDocument();
        expect(screen.queryByLabelText('coordinator@42.nl')).toBeInTheDocument();
        expect(screen.queryByLabelText('user@42.nl')).not.toBeInTheDocument();
      });
    });

    it('should call onChange with custom typeahead option when allowNew is true and option does not exist', async () => {
      expect.assertions(9);

      const { onChangeSpy } = setup({
        allowNew: true
      });

      fireEvent.change(screen.getByRole('textbox'), { target: { value: 'Test' } });

      await waitFor(() => {
        expect(screen.queryByLabelText('Test')).toBeInTheDocument();
      });

      fireEvent.click(screen.getByLabelText('Test'));

      expect(onChangeSpy).toHaveBeenCalledTimes(1);
      expect(onChangeSpy.mock.calls[0][0]).toMatchObject([{ label: 'Test', customOption: true }]);
    });

    describe('value changes', () => {
      test('becomes empty', () => {
        const { props, rerender } = setup({
          value: [ adminUser() ]
        });

        expect(screen.queryByText('admin@42.nl', { exact: false })).toBeInTheDocument();

        const newProps = {
          ...props,
          value: undefined
        };

        rerender(
          <TypeaheadMultiple {...newProps} />
        );

        expect(screen.queryByText('admin@42.nl', { exact: false })).not.toBeInTheDocument();
      });

      test('becomes filled', () => {
        const { props, rerender } = setup({});

        expect(screen.queryByText('admin@42.nl', { exact: false })).not.toBeInTheDocument();

        const newProps = {
          ...props,
          value: [ adminUser() ]
        };

        rerender(
          <TypeaheadMultiple {...newProps} />
        );

        expect(screen.queryByText('admin@42.nl', { exact: false })).toBeInTheDocument();
      });
    });
  });
});
