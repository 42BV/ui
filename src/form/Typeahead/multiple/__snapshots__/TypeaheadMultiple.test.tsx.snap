// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: TypeaheadMultiple ui with value: Component: TypeaheadMultiple => ui => with value 1`] = `
<FormGroup
  className=""
  tag="div"
>
  <Label
    for="bestFriend"
    tag="label"
    widths={
      Array [
        "xs",
        "sm",
        "md",
        "lg",
        "xl",
      ]
    }
  >
    Best friend
  </Label>
  <div
    className=""
  >
    <asyncContainer(TypeaheadComponent)
      delay={200}
      id="bestFriend"
      inputProps={
        Object {
          "className": "form-control",
          "value": Array [
            Object {
              "active": true,
              "email": "admin@42.nl",
              "id": 42,
              "roles": Array [
                "ADMIN",
              ],
            },
          ],
        }
      }
      isLoading={false}
      minLength={2}
      multiple={true}
      onChange={[Function]}
      onSearch={[Function]}
      options={Array []}
      placeholder="Please provide your best friend"
      promptText="Type to search..."
      searchText="Searching..."
      selected={
        Array [
          Object {
            "label": "admin@42.nl",
            "value": Object {
              "active": true,
              "email": "admin@42.nl",
              "id": 42,
              "roles": Array [
                "ADMIN",
              ],
            },
          },
        ]
      }
      useCache={true}
    />
  </div>
  Some error
</FormGroup>
`;

exports[`Component: TypeaheadMultiple ui without placeholder: Component: TypeaheadMultiple => ui => without placeholder 1`] = `
<FormGroup
  className=""
  tag="div"
>
  <Label
    for="bestFriend"
    tag="label"
    widths={
      Array [
        "xs",
        "sm",
        "md",
        "lg",
        "xl",
      ]
    }
  >
    Best friend
  </Label>
  <div
    className=""
  >
    <asyncContainer(TypeaheadComponent)
      delay={200}
      id="bestFriend"
      inputProps={
        Object {
          "className": "form-control",
          "value": Array [
            Object {
              "active": true,
              "email": "admin@42.nl",
              "id": 42,
              "roles": Array [
                "ADMIN",
              ],
            },
          ],
        }
      }
      isLoading={false}
      minLength={2}
      multiple={true}
      onChange={[Function]}
      onSearch={[Function]}
      options={Array []}
      promptText="Type to search..."
      searchText="Searching..."
      selected={
        Array [
          Object {
            "label": "admin@42.nl",
            "value": Object {
              "active": true,
              "email": "admin@42.nl",
              "id": 42,
              "roles": Array [
                "ADMIN",
              ],
            },
          },
        ]
      }
      useCache={true}
    />
  </div>
  Some error
</FormGroup>
`;
